sdks/wasm/bindings.cs

```diff
@@ -0,0 +1,174 @@
+
+		public static string InvokeJS (string str)
+		{
+			int exception = 0;
+			var res = InvokeJS (str, out exception);
+			if (exception != 0)
+				throw new JSException (res);
+			return res;
+		}
+
+		static Dictionary<int, JSObject> bound_objects = new Dictionary<int, JSObject>();
+		static Dictionary<object, JSObject> raw_to_js = new Dictionary<object, JSObject>();
+
+		static int BindJSObject (int js_id) {
+			JSObject obj;
+			if (bound_objects.ContainsKey (js_id))
+				obj = bound_objects [js_id];
+			else
+				bound_objects [js_id] = obj = new JSObject (js_id);
+
+			return (int)(IntPtr)obj.Handle;
+		}
+
+		static object CreateTaskSource (int js_id) {
+			return new TaskCompletionSource<object> ();
+		}
+
+		static void SetTaskSourceResult (TaskCompletionSource<object> tcs, object result) {
+			tcs.SetResult (result);
+		}
+
+		static void SetTaskSourceFailure (TaskCompletionSource<object> tcs, string reason) {
+			tcs.SetException (new JSException (reason));
+		}
+
+		static int GetTaskAndBind (TaskCompletionSource<object> tcs, int js_id) {
+			return BindExistingObject (tcs.Task, js_id);
+		}
+
+		static int BindExistingObject (object raw_obj, int js_id) {
+			JSObject obj;
+			if (raw_obj is JSObject)
+				obj =(JSObject)raw_obj;
+			else if (raw_to_js.ContainsKey (raw_obj))
+				obj = raw_to_js [raw_obj];
+			else
+				raw_to_js [raw_obj] = obj = new JSObject (js_id, raw_obj);
+
+			return (int)(IntPtr)obj.Handle;
+		}
+
+		static int GetJSObjectId (object raw_obj) {
+			JSObject obj = null;
+			if (raw_obj is JSObject)
+				obj =(JSObject)raw_obj;
+			else if (raw_to_js.ContainsKey (raw_obj))
+				obj = raw_to_js [raw_obj];
+
+			var js_handle = obj != null ? obj.JSHandle : -1;
+
+			return js_handle;
+		}
+
+		static object GetMonoObject (int gc_handle) {
+			GCHandle h = (GCHandle)(IntPtr)gc_handle;
+			JSObject o = (JSObject)h.Target;
+			if (o != null && o.RawObject != null)
+				return o.RawObject;
+			return o;
+		}
+
+		static object BoxInt (int i) {
+			return i;
+		}
+		static object BoxDouble (double d) {
+			return d;
+		}
+
+		static object BoxBool (int b) {
+			return b == 0 ? false : true;
+		}
+
+		static MethodInfo gsjsc;
+		static void GenericSetupJSContinuation<T> (Task<T> task, JSObject cont_obj) {
+			task.GetAwaiter().OnCompleted (() => {
+				//FIXME we should dispose cont_obj after completing the Promise
+				if (task.Exception != null)
+					cont_obj.Invoke ("reject", task.Exception.ToString ());
+				else {
+					cont_obj.Invoke ("resolve", task.Result);
+				}
+			});
+		}
+
+		static void SetupJSContinuation (Task task, JSObject cont_obj) {
+			if (task.GetType () == typeof (Task)) {
+				task.GetAwaiter().OnCompleted (() => {
+					//FIXME we should dispose cont_obj after completing the Promise
+					if (task.Exception != null)
+						cont_obj.Invoke ("reject", task.Exception.ToString ());
+					else
+						cont_obj.Invoke ("resolve", null);
+				});
+			} else {
```

Comment:
Could you comment what you expect this to be when `task.GetType () != typeof(Task)` ?

Reply:
It's going to be `Task<T>`.

Notice that `task` is of type `Task` so the only possible outcome of `task.GetType () != typeof (Task)` is a subclass of it.

What will happen in the case of a different subclass is that we'll throw an exception at runtime.